File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2021/Y21D21.kt
package org.gristle.pdxpuzzles.advent.y2021

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.iteration.max

class Y21D21(input: String) : Day {

    class DeterministicDie {
        private var die = 0

        var counter = 0L

        fun roll(n: Int): List<Int> {
            counter += n
            return (1..n).fold(listOf()) { acc, _ ->
                die = if (die == 100) 1 else die + 1
                acc + die
            }
        }
    }

    class Game(private val spaces: Int, private val die: DeterministicDie = DeterministicDie()) {

        data class GameResult(val winner: String, val winScore: Long, val loseScore: Long, val dieRolls: Long) {
            fun solve() = loseScore * dieRolls
        }

        private fun Int.advance(n: Int) = (this - 1 + n) % spaces + 1

        fun play(p1Start: Int, p2Start: Int, winCondition: Long = 1000L): GameResult {
            var p1Score = 0L
            var p2Score = 0L
            var p1Pos = p1Start
            var p2Pos = p2Start

            while (p2Score < winCondition) {
                val p1Roll = die.roll(3)
                p1Pos = p1Pos.advance(p1Roll.sum())
                p1Score += p1Pos
//                println("Player 1 rolls $p1Roll and moves to space $p1Pos for a total score of $p1Score.")
                if (p1Score >= winCondition) {
                    return GameResult("p1", p1Score, p2Score, die.counter)
                }
                val p2Roll = die.roll(3)
                p2Pos = p2Pos.advance(p2Roll.sum())
                p2Score += p2Pos
//                println("Player 2 rolls $p2Roll and moves to space $p2Pos for a total score of $p2Score.")
            }
            return GameResult("p2", p2Score, p1Score, die.counter)
        }
    }

    private val matchValues = """(\d+)${'$'}"""
        .toRegex(RegexOption.MULTILINE)
        .findAll(input)
        .map { it.value.toInt() }

    private val p1Start = matchValues.first()
    private val p2Start = matchValues.last()

    override fun part1(): Long {
        val game = Game(10)
        return game.play(p1Start, p2Start).solve()
    }

    private val rf = listOf(3 to 1, 4 to 3, 5 to 6, 6 to 7, 7 to 6, 8 to 3, 9 to 1)

    private fun wins(p1: Int, t1: Int, p2: Int, t2: Int): Pair<Long, Long> {
        if (t2 <= 0) return 0L to 1L

        var w1 = 0L
        var w2 = 0L
        for ((roll, frequency) in rf) {
            val (c2, c1) = wins(p2, t2, (p1 + roll) % 10, t1 - 1 - (p1 + roll) % 10)
            w1 += frequency * c1
            w2 += frequency * c2
        }
        return w1 to w2
    }

    override fun part2() = wins(p1Start - 1, 21, p2Start - 1, 21).max()
}

fun main() = Day.runDay(Y21D21::class)

//    Class creation: 17ms
//    Part 1: 605070 (2ms)
//    Part 2: 218433063958910 (1004ms)
//    Total time: 1024ms

File: rust/advent/src/bin/2021/y21d21.rs
use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input = (usize, usize);
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(21, 21).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

struct Die {
    die: usize,
    counter: usize,
}

impl Die {
    fn roll(&mut self, n: usize) -> usize {
        self.counter += n;
        (0..n).fold(0, |acc, _| {
            self.die = if self.die == 100 { 1 } else { self.die + 1 };
            acc + self.die
        })
    }
}


fn parse_input(input: &str) -> Input {
    let (_, p1, _, p2) = input.get_numbers().collect_tuple().unwrap();
    (p1, p2)
}

fn part1((p1_start, p2_start): &Input) -> Output {
    let advance = |p: usize, n: usize| (p - 1 + n) % 10 + 1;
    let mut p1_score = 0;
    let mut p2_score = 0;
    let mut p1_pos = *p1_start;
    let mut p2_pos = *p2_start;
    let mut die = Die { die: 0, counter: 0 };

    loop {
        let p1_roll = die.roll(3);
        p1_pos = advance(p1_pos, p1_roll);
        p1_score += p1_pos;
        // println!("Player 1 rolls {p1_roll} and moves to space {p1_pos} for a total score of {p1_score}.");
        if p1_score >= 1_000 {
            return die.counter * p2_score;
        }
        let p2_roll = die.roll(3);
        p2_pos = advance(p2_pos, p2_roll);
        p2_score += p2_pos;
        // println!("Player 2 rolls {p2_roll} and moves to space {p2_pos} for a total score of {p2_score}.");
        if p2_score >= 1_000 {
            return die.counter * p1_score;
        }
    }
}

fn part2((p1_start, p2_start): &Input) -> Output {
    let rf = [(3, 1), (4, 3), (5, 6), (6, 7), (7, 6), (8, 3), (9, 1)];

    fn wins(p1: i64, t1: i64, p2: i64, t2: i64, rf: &[(i64, i64)]) -> (i64, i64) {
        if t2 <= 0 {
            return (0, 1);
        }

        let mut w1 = 0;
        let mut w2 = 0;
        for &(roll, frequency) in rf {
            let (c2, c1) = wins(p2, t2, (p1 + roll) % 10, t1 - 1 - (p1 + roll) % 10, rf);
            w1 += frequency * c1;
            w2 += frequency * c2;
        }
        (w1, w2)
    }
    let (w1, w2) = wins(
        *p1_start as i64 - 1,
        21,
        *p2_start as i64 - 1,
        21,
        &rf
    );
    if w1 > w2 { w1 as usize } else { w2 as usize }
}

#[test]
fn default() {
    let input = get_input(21, 21).unwrap();
    let input = parse_input(&input);
    assert_eq!(605070, part1(&input));
    assert_eq!(218433063958910, part2(&input));
}

// Input parsed (19μs)
// 1. 605070 (8μs)
// 2. 218433063958910 (190ms)
// Total: 190ms